scripts/tasks/preprocess-api-history.ts

```diff
@@ -0,0 +1,198 @@
+import globby from 'globby';
+import path from 'path';
+import { fromMarkdown } from 'mdast-util-from-markdown';
+import { visit } from 'unist-util-visit';
+import type { Html, Root } from 'mdast';
+import type { Node, Literal } from 'unist';
+import Ajv, { JSONSchemaType, ValidateFunction } from 'ajv';
+import { readFile, writeFile } from 'fs/promises';
+import { fromHtml } from 'hast-util-from-html';
+import logger from '@docusaurus/logger';
+import { parse as parseYaml } from 'yaml';
+
+const apiHistoryRegex = /<!--[\s\S]*?(```[\s\S]*?```)[\s\S]*?-->/;
+
+interface ChangeSchema {
+  'pr-url': string;
+  'breaking-changes-header'?: string;
+  description?: string;
+}
+
+interface ApiHistory {
+  added?: ChangeSchema[];
+  deprecated?: ChangeSchema[];
+  changes?: ChangeSchema[];
+}
+
+interface LiteralString extends Literal {
+  value: string;
+}
+
+// Copied from here: <https://github.com/electron/website/blob/feat/api-history/scripts/tasks/add-frontmatter.ts#L16-L23>
+const getMarkdownFiles = async (startPath: string) => {
+  const filesPaths = await globby(path.posix.join(startPath, '**/*.md'));
+
+  const files: Map<string, string> = new Map();
+  for (const filePath of filesPaths) {
+    const content = await readFile(filePath, 'utf-8');
+    files.set(filePath, content);
+  }
+
+  return files;
```

Comment:
Use concurrency to achieve better performance.
```suggestion
  const filePromises: Promise<[string, string]>[] = filesPaths.map(f => {
    return fs.readFile(f, "utf-8").then(content => [f, content]);
  });
  
  const contents = await Promise.all(filePromises);
  return new Map(contents);
```

Reply:
I measured the performance for that suggestion and got a significantly slower result:

```
238 Markdown documents
~30ms		current
100-300ms	suggestion
```

We also decided to reduce the number of Markdown files we check in [`247b7d9` (#594)](https://github.com/electron/website/pull/594/commits/247b7d964e309538371c0a26631061832580ed23). Same trend in that scenario:

```
70 Markdown documents
~10ms		current
40-100ms	suggestion
```

Reply:
Oh, that's interesting, using concurrency actually made it slower.😅🤔

Reply:
I just tried it locally, and the result was the opposite of yours. Can you double-check? Here is my test code:

```js
import globby from 'globby';
import path from 'path';
import { readFile } from 'fs/promises';

const startPath = 'docs/latest';

const filesPaths = await globby(path.posix.join(startPath, 'api', '/*.md'));

console.log('count:', filesPaths.length);

const getMarkdownFiles = async () => {
  const files = new Map();
  for (const filePath of filesPaths) {
    const content = await readFile(filePath, 'utf-8');
    files.set(filePath, content);
  }

  return files;
};

const getMarkdownFiles2 = async () => {
  const ps = filesPaths.map((f) => {
    return readFile(f, 'utf-8').then((content) => [f, content]);
  });

  const contents = await Promise.all(ps);
  return new Map(contents);
};

console.time('getMarkdownFiles1');
await getMarkdownFiles();
console.timeEnd('getMarkdownFiles1');

console.time('getMarkdownFiles2');
await getMarkdownFiles2();
console.timeEnd('getMarkdownFiles2');
```

Result:

```
$ node ./test.mjs
count: 72
getMarkdownFiles1: 12.383ms
getMarkdownFiles2: 5.1ms
```

My machine info: `MacBook Pro Intel Core i9 2.4GHz 32GB`

Reply:
Hmm, I tried your test code and got the same result as you. I'm probably just using the profiler wrong then:

```patch
diff --git a/scripts/tasks/preprocess-api-history.ts b/scripts/tasks/preprocess-api-history.ts
index fb95451..5cbc3af 100644
--- a/scripts/tasks/preprocess-api-history.ts
+++ b/scripts/tasks/preprocess-api-history.ts
@@ -33,13 +33,12 @@ interface LiteralString extends Literal {
 const getMarkdownFiles = async (startPath: string) => {
   const filesPaths = await globby(path.posix.join(startPath, 'api', '/*.md'));
 
-  const files: Map<string, string> = new Map();
-  for (const filePath of filesPaths) {
-    const content = await readFile(filePath, 'utf-8');
-    files.set(filePath, content);
-  }
+  const filePromises: Promise<[string, string]>[] = filesPaths.map(f => {
+    return readFile(f, 'utf-8').then((content) => [f, content]);
+  });
 
-  return files;
+  const contents = await Promise.all(filePromises);
+  return new Map(contents);
 };
 
 // Copied from here but slightly different: <https://github.com/electron/lint-roller/blob/bac245478ba69017c2a82e8ba1b2884a37647494/bin/lint-markdown-api-history.ts#L42-L67>
@@ -166,7 +165,9 @@ export const preprocessApiHistory = async (startPath: string) => {
     return;
   }
 
+  console.profile('preprocessApiHistory');
   const files = await getMarkdownFiles(startPath);
+  console.profileEnd('preprocessApiHistory');
 
   for (const [filePath, content] of files) {
     if (!content.includes('<!--')) continue;

```

| Original | Suggestion |
|:---------:|:------------:|
| ![Screenshot 2024-07-16 154133](https://github.com/user-attachments/assets/c31ecdf3-3697-46b7-8a1c-f74e4b959258) | ![Screenshot 2024-07-16 154416](https://github.com/user-attachments/assets/c3f76911-7505-4696-9806-40cf41f71bb9) |

Reply:
If my profiling is wrong and you're right though, it's only like a `~5ms` difference. Maybe at that point it's better to use the current code since it's easier to debug? @dsanders11

Reply:
Let's stick with the current code. It was a good suggestion from @BlackHole1 to try out, but given the mixed results and minor potential difference, I don't think it's worth pursuing here.

Reply:
OK, I can accept both:)